### ISSUE TITLE
Meta Person Loader Error with exporting GLTF and ZIPs. Only unzipped GLB works well.

### ISSUE THREAD

**Issue by Andrey-Kachow** [avatarsdk/metaperson-loader-unity#7]
## Meta Person Loader Error with exporting GLTF and Zip. GLB is fine.

**TLDR** - _**The Table at the bottom shows the mapping from SceneHandler.modelUrl to Error Message**_

When working with the Unity Loader Scene, I struggled to make the scene work for me when I tried to export my models in different formats.
The project README claims that the link can be submitted in **either** GLB or GLTF and ZIP formats.
However, I was successful with only unzipped GLB.
The models were generated with the Customised MetaPersonCreator with various (hardcoded) export parameters.
They were then uploaded to my website and made available by the route entry points snippet, which is summarised in the snippet below. This snippet provides a detailed overview of the folder structure and expands the zip archives.

```

http://satprotocol.online/instance/avatars
│
└── debug
    ├── glb__modelka
    │   └── model.glb
    │
    ├── gltf_unzipped__modelechka
    │   ├── model.bin
    │   ├── AvatarTeeth_Normal1_1K.jpg
    │   ├── AvatarHeadMale_Normal1_1K.jpg
    │   ├── AvatarTeeth_GltfMetallicRoughness_1K.jpg
    │   ├── model.gltf
    │   ├── AvatarBodyMale_Normal1_1K.jpg
    │   ├── AvatarBodyMale_Color_1K.jpg
    │   ├── AvatarHeadMale_Color_1K.jpg
    │   ├── AvatarTeeth_Color_1K.jpg
    │   ├── AvatarEyes_Color_512.jpg
    │   ├── AvatarEyes_Normal_512.jpg
    │   ├── AvatarRightCornea_Color_512.jpg
    │   ├── KARI_Normal_1K.jpg
    │   ├── KARI_GltfMetallicRoughness_1K.jpg
    │   ├── AvatarHeadMale_GltfMetallicRoughness_1K.jpg
    │   ├── KARI_Color_1K.jpg
    │   ├── AvatarLeftCornea_Color_512.jpg
    │   ├── AvatarEyes_GltfMetallicRoughness_512.jpg
    │   └── AvatarBodyMale_GltfMetallicRoughness_1K.jpg
    │
    ├── gltf_zipped__modelushechka
    │   └── avatar_gltf.zip (expanded)
    │       ├── model.bin
    │       ├── AvatarTeeth_Normal1_1K.jpg
    │       ├── AvatarHeadMale_Normal1_1K.jpg
    │       ├── AvatarTeeth_GltfMetallicRoughness_1K.jpg
    │       ├── model.gltf
    │       ├── AvatarBodyMale_Normal1_1K.jpg
    │       ├── AvatarBodyMale_Color_1K.jpg
    │       ├── AvatarHeadMale_Color_1K.jpg
    │       ├── AvatarTeeth_Color_1K.jpg
    │       ├── AvatarEyes_Color_512.jpg
    │       ├── AvatarEyes_Normal_512.jpg
    │       ├── AvatarRightCornea_Color_512.jpg
    │       ├── KARI_Normal_1K.jpg
    │       ├── KARI_GltfMetallicRoughness_1K.jpg
    │       ├── AvatarHeadMale_GltfMetallicRoughness_1K.jpg
    │       ├── KARI_Color_1K.jpg
    │       ├── AvatarLeftCornea_Color_512.jpg
    │       ├── AvatarEyes_GltfMetallicRoughness_512.jpg
    │       └── AvatarBodyMale_GltfMetallicRoughness_1K.jpg
    │
    └── glb_zipped__Modelishche
        └── avatar_glb.zip (expanded)
            └── model.glb
```

Various URL Parameters were attempted in the `SceneHadnler.modelUrl`; only 1 out of 6 succeeded.
The table below provides the relationship between the URL and the error message generated by Unity. Some Cases logged more than one log message. For the sake of being concise, only such a log per case is included.
The Scene was played only within the Unity.
No proper WebGL build was made (Build and Run).

<table>
  <tr>
    <th>SceneHandler <br> URL</th>
    <th>Success/Error log in Unity</th>
  </tr>
  <tr>
    <td>
        <a href="http://satprotocol.online/instance/avatars/debug/gltf_zipped__modelushechka/avatar_gltf.zip">
            http:// <br> satprotocol.online <br> /instance/avatars/ <br> debug/ <br> gltf_zipped_ <br> _modelushechka/ <br> avatar_gltf.zip 
        </a>
    </td>
    <td>
        <code>
JsonParsingFailed
UnityEngine.Debug:LogError (object)
GLTFast.GltfImportBase/<ParseJsonAndLoadBuffers>d__101:MoveNext () (at Library/PackageCache/com.unity.cloud.gltfast@6.0.1/Runtime/Scripts/GltfImport.cs:1056)
UnityEngine.UnitySynchronizationContext:ExecuteTasks ()
        </code>
    </td>
  </tr>
  <tr>
    <td>
        <a href="http://satprotocol.online/instance/avatars/debug/gltf_unzipped__modelechka/model.gltf">
            http:// <br> satprotocol.online <br> /instance/avatars/ <br> debug/ <br> gltf_unzipped_ <br> _modelechka/ <br> model.gltf
        </a>
    </td>
    <td>
        <code>
Exception during avatar loading: URI must be absolute
UnityEngine.Debug:LogErrorFormat (string,object[])
AvatarSDK.MetaPerson.Loader.MetaPersonLoader/<LoadModelAsync>d__7:MoveNext () (at Library/PackageCache/com.avatarsdk.metaperson.loader@bacac4e357/Runtime/Scripts/MetaPersonLoader.cs:104)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<byte[]>:SetResult (byte[])
AvatarSDK.MetaPerson.Loader.MetaPersonLoader/<DownloadFileAsync>d__9:MoveNext () (at Library/PackageCache/com.avatarsdk.metaperson.loader@bacac4e357/Runtime/Scripts/MetaPersonLoader.cs:134)
UnityEngine.UnitySynchronizationContext:ExecuteTasks ()
        </code>
    </td>
  </tr>
  <tr>
    <td>
        <a href="http://satprotocol.online/instance/avatars/debug/gltf_unzipped__modelechka/model.bin">
            http:// <br> satprotocol.online <br>
            /instance/avatars/ <br> debug/ <br> gltf_unzipped_ <br> _modelechka/ <br>
            model.bin
        </a>
    </td>
    <td>
        <code>
Exception during avatar loading: Object reference not set to an instance of an object
UnityEngine.Debug:LogErrorFormat (string,object[])
AvatarSDK.MetaPerson.Loader.MetaPersonLoader/<LoadModelAsync>d__7:MoveNext () (at Library/PackageCache/com.avatarsdk.metaperson.loader@bacac4e357/Runtime/Scripts/MetaPersonLoader.cs:104)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<bool>:SetResult (bool)
GLTFast.GltfImportBase/<Load>d__59:MoveNext () (at Library/PackageCache/com.unity.cloud.gltfast@6.0.1/Runtime/Scripts/GltfImport.cs:457)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<bool>:SetResult (bool)
GLTFast.GltfImportBase/<LoadGltfJson>d__62:MoveNext () (at Library/PackageCache/com.unity.cloud.gltfast@6.0.1/Runtime/Scripts/GltfImport.cs:568)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<bool>:SetResult (bool)
GLTFast.GltfImportBase/<LoadGltf>d__104:MoveNext () (at Library/PackageCache/com.unity.cloud.gltfast@6.0.1/Runtime/Scripts/GltfImport.cs:1177)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<bool>:SetResult (bool)
GLTFast.GltfImportBase/<ParseJsonAndLoadBuffers>d__101:MoveNext () (at Library/PackageCache/com.unity.cloud.gltfast@6.0.1/Runtime/Scripts/GltfImport.cs:1104)
UnityEngine.UnitySynchronizationContext:ExecuteTasks ()
        </code>
    </td>
  </tr>
  <tr>
    <td>
        <span>
            (directory`*`) <br>
            http:// <br> satprotocol.online <br> /instance/avatars/ <br> debug/ <br> gltf_unzipped_ <br> _modelechka/
        </span>
    </td>
    <td>
        <code>
JsonParsingFailed
UnityEngine.Debug:LogError (object)
GLTFast.GltfImportBase/<ParseJsonAndLoadBuffers>d__101:MoveNext () (at Library/PackageCache/com.unity.cloud.gltfast@6.0.1/Runtime/Scripts/GltfImport.cs:1056)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<bool>:Start<GLTFast.GltfImportBase/<ParseJsonAndLoadBuffers>d__101> (GLTFast.GltfImportBase/<ParseJsonAndLoadBuffers>d__101&)
GLTFast.GltfImportBase:ParseJsonAndLoadBuffers (string,System.Uri)
GLTFast.GltfImportBase/<LoadGltf>d__104:MoveNext () (at Library/PackageCache/com.unity.cloud.gltfast@6.0.1/Runtime/Scripts/GltfImport.cs:1174)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<bool>:Start<GLTFast.GltfImportBase/<LoadGltf>d__104> (GLTFast.GltfImportBase/<LoadGltf>d__104&)
GLTFast.GltfImportBase:LoadGltf (string,System.Uri)
GLTFast.GltfImportBase/<LoadGltfJson>d__62:MoveNext () (at Library/PackageCache/com.unity.cloud.gltfast@6.0.1/Runtime/Scripts/GltfImport.cs:561)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<bool>:Start<GLTFast.GltfImportBase/<LoadGltfJson>d__62> (GLTFast.GltfImportBase/<LoadGltfJson>d__62&)
GLTFast.GltfImportBase:LoadGltfJson (string,System.Uri,GLTFast.ImportSettings,System.Threading.CancellationToken)
GLTFast.GltfImportBase/<Load>d__59:MoveNext () (at Library/PackageCache/com.unity.cloud.gltfast@6.0.1/Runtime/Scripts/GltfImport.cs:456)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<bool>:Start<GLTFast.GltfImportBase/<Load>d__59> (GLTFast.GltfImportBase/<Load>d__59&)
GLTFast.GltfImportBase:Load (byte[],System.Uri,GLTFast.ImportSettings,System.Threading.CancellationToken)
AvatarSDK.MetaPerson.Loader.MetaPersonLoader/<LoadModelAsync>d__7:MoveNext () (at Library/PackageCache/com.avatarsdk.metaperson.loader@bacac4e357/Runtime/Scripts/MetaPersonLoader.cs:71)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<byte[]>:SetResult (byte[])
AvatarSDK.MetaPerson.Loader.MetaPersonLoader/<DownloadFileAsync>d__9:MoveNext () (at Library/PackageCache/com.avatarsdk.metaperson.loader@bacac4e357/Runtime/Scripts/MetaPersonLoader.cs:134)
UnityEngine.UnitySynchronizationContext:ExecuteTasks ()
        </code>
    </td>
  </tr>
  <tr>
    <td>
        <a href="http://satprotocol.online/instance/avatars/debug/glb_zipped__Modelishche/avatar_glb.zip">
            http:// <br> satprotocol.online <br> /instance/avatars/ <br> debug/ <br> glb_zipped_ <br> _Modelishche/ <br> avatar_glb.zip
        </a>
    </td>
    <td>
        <code>
Exception during avatar loading: Value cannot be null.
Parameter name: value
UnityEngine.Debug:LogErrorFormat (string,object[])
AvatarSDK.MetaPerson.Loader.MetaPersonLoader/<LoadModelAsync>d__7:MoveNext () (at Library/PackageCache/com.avatarsdk.metaperson.loader@bacac4e357/Runtime/Scripts/MetaPersonLoader.cs:104)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<byte[]>:SetResult (byte[])
AvatarSDK.MetaPerson.Loader.MetaPersonLoader/<DownloadFileAsync>d__9:MoveNext () (at Library/PackageCache/com.avatarsdk.metaperson.loader@bacac4e357/Runtime/Scripts/MetaPersonLoader.cs:134)
UnityEngine.UnitySynchronizationContext:ExecuteTasks ()
        </code>
    </td>
  </tr>
  <tr>
    <td>
        <a href="http://satprotocol.online/instance/avatars/debug/glb__modelka/model.glb">
            http:// <br> satprotocol.online <br> /instance/avatars/ <br> debug/ <br> glb__modelka/ <br> model.glb
        </a>
    </td>
    <td>
        &#128077;
    </td>
  </tr>
</table>

`*`The server is not configured to serve subdirectories of `/instance` directory.
Only files can be accessed

I hope that was useful.

**Comment by dledents** [comment]
I was able to reproduce this issue only with the zipped GLTF model. The following export parameters were used. 
```
let exportParametersMessage = {
    "eventName": "set_export_parameters",
    "format" : "gltf",
    "lod" : 1,
    "textureProfile" : "1K.png",
    "useZip" : true
};
evt.source.postMessage(exportParametersMessage, "*");
```

I can't reproduce it with the zipped GLB model. If you have such a file, please send it. The provided links doesn't work because they require to log in.

**Comment by Andrey-Kachow** [comment]
Sure, I'm sorry for any trouble. Authentication was recently added to my web app, preventing you from accessing many assets. The zip archive repeats the files in the hierarchy and is attached to this comment. My issues may have come from a misunderstanding of the intended usage. Now, when I learned about avatar storage already provided, using the **`metaperson-loader-unity`** is a pleasure. I hope my contributions are more helpful than they are disruptive ;)

[debug.zip](https://github.com/user-attachments/files/15764221/debug.zip)

**Comment by dledents** [comment]
The issue should be fixed in the latest commit in master. @Andrey-Kachow, could you check if it works for you?

**Comment by Andrey-Kachow** [comment]
Thank you for your effort. Here is the update. Do you expect uncompressed GLTF to work? Apart from that, it either works appropriately or seems irrelevant now that I have a better understanding.

| Suffix       |  Result  |
|----------------|----------------|
| `*.glb`   | :+1:   |
| `*.zip`   | :+1:  |
| `*.gltf`   | :x:    |
| `*.bin`  | n/a  |
| `gltf` dir   | n/a  |

It now works well for all ZIP cases. I get this error for the uncompressed GLTF case. (debug link is now available without signing in [http://satprotocol.online/instance/avatars/debug/gltf_unzipped__modelechka/model.gltf](http://satprotocol.online/instance/avatars/debug/gltf_unzipped__modelechka/model.gltf) )

```
Exception during avatar loading: Object reference not set to an instance of an object
UnityEngine.Debug:LogErrorFormat (string,object[])
AvatarSDK.MetaPerson.Loader.MetaPersonLoader/<LoadModelAsync>d__7:MoveNext () (at Library/PackageCache/com.avatarsdk.metaperson.loader@c12a588ae0/Runtime/Scripts/MetaPersonLoader.cs:120)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<bool>:SetException (System.Exception)
GLTFast.GltfImportBase/<Load>d__59:MoveNext () (at Library/PackageCache/com.unity.cloud.gltfast@6.0.1/Runtime/Scripts/GltfImport.cs:456)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<bool>:SetException (System.Exception)
GLTFast.GltfImportBase/<LoadGltfJson>d__62:MoveNext () (at Library/PackageCache/com.unity.cloud.gltfast@6.0.1/Runtime/Scripts/GltfImport.cs:567)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<bool>:SetException (System.Exception)
GLTFast.GltfImportBase/<Prepare>d__118:MoveNext () (at Library/PackageCache/com.unity.cloud.gltfast@6.0.1/Runtime/Scripts/GltfImport.cs:2173)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder:SetResult ()
GLTFast.TimeBudgetPerFrameDeferAgent/<BreakPoint>d__11:MoveNext () (at Library/PackageCache/com.unity.cloud.gltfast@6.0.1/Runtime/Scripts/TimeBudgetPerFrameDeferAgent.cs:85)
UnityEngine.UnitySynchronizationContext:ExecuteTasks ()
```
After a few seconds, I also got this dodgy log message that I could not expand even with full stack trace in player settings.
```
A Native Collection has not been disposed, resulting in a memory leak. Enable Full StackTraces to get more details.
```

**Comment by dledents** [comment]
The uncompressed GLTF doesn't work if URI points to a network resource. It is expected behavior because the GLTF contains relative paths to textures. If the URI points to a local GLTF file, the model should be loaded.
When the URI points to a BIN file or directory, it is considered as an invalid and the model isn't loaded.

**Comment by dledents** [comment]
> A Native Collection has not been disposed, resulting in a memory leak. Enable Full StackTraces to get more details.
Never mind this warning. It doesn't affect the plugin's functionality.

**Comment by Andrey-Kachow** [comment]
Thank you for replying. If the network-located GLTF support is not intended, then the issue can be closed as both GLB and ZIP archived variants of GLB and GLTF work now. Thank you for your effort, and good luck with the project.
Best,
Andrey
